enabled_if

rgrinberg opened the pull request:
# Problem

The problem this PR is trying to solve is a bit specific to dune, but it's likely that it will be encountered by other projects as well. Currently, dune has a complicated script for generating test definitions that make sure tests run only in environments which support them. For example, some tests may not run on win32. A mechanism to enable tests conditionally would fix these problems.

# Proposal

Add an `enabled_if` field to aliases to toggle the execution of the alias. This field will be valued by a little EDSL for expressing boolean expressions. Here's an example of the kind of conditions we'd express with it:

```
(alias
 ((name runtest)
  (deps (foo.exe))
  (action (run ${<}))
  (enabled_if (and (<> ${os} win32) (>= ${ocaml_version} 4.0.5)))))
```

# Progress

This stalled a bit since I'm not sure how to do handle type safety here. Ideally, we only allow numbers and versions to be compared with `>=`, `<`, etc. I guess we really need to annotate which variables are comparable and making sure that we don't compare numbers to strings or versions.

diml commented:
this boolean language looks quite simple and during development all expressions should be evaluated on every run, so runtime checks will do.

Currently we already keep the type of variable expansion (string or path). We can add a version type

rgrinberg commented:
Okay, makes sense. I'm wondering how we should organize things however. `Blang` could certainly live in its own module, but `Var_expansion.t` isn't really specific to actions anymore.

diml commented:
Maybe a new `Value` module? That's basically what it represents. BTW, let's do https://github.com/ocaml/dune/issues/701#issuecomment-381939574 first as it formalizes the interpretation of variables.

rgrinberg commented:
@diml now that we have a Value module, I'd like to move this along. One question is how to handle expressions like : `(enabled_if %{ocaml_version})`. In other words, what happens if we try to use a non boolean valued expression as the condition in `enabled_if`. Just a runtime error?

diml commented:
Yes, a runtime error seems fine for now. These expressions are very simple and will be evaluated often during development, so there is no need for heavy machinery.

rgrinberg commented:
Okay. One last point is that we might be closing the door on things like `(enabled_if %{read:file-with-flag})` since there would be no way to read a file into a boolean like this. This is a bit awkward, but I can imagine wanting a toggle via env var for example. In which case, an awkward form would have to be invented: `(enabled_if %{env-var-set-p:FOO})`

diml commented:
That seems fine, but if we add support for reading environment variables like this, then `(setenv X foo (echo %{env:X})` must print `"foo"`.

rgrinberg commented:
Okay, this PR has been revived and seems to be in a good enough form for a v1 of the feature. The only thing that remains is to make sure that the field is available only for `1.1` version of the language. The current version doesn't support advanced features like `%{read:..}` stanzas, but should be good enough for our use case in dune.

rgrinberg commented:
I added support for the `(1, 1)` version guard.

rgrinberg commented:
This PR is a pre-req as well for #924.

diml commented:
Ok, could you write the documentation for this feature? I'd like to read the semantic of boolean expressions before reading the code that implements it.

diml commented:
Thanks, that seems reasonable

diml commented on src/blang.ml:
```diff
@@ -0,0 +1,56 @@
+  | Or of 'a t list
+  | Compare of Op.t * 'a * 'a
+
+let compare_vals ~dir (x : Value.t) (y : Value.t) =
```
What about putting this in `Value.L`?

diml commented on src/simple_rules.ml:
```diff
@@ -79,23 +79,37 @@ let alias sctx ~dir ~scope (alias_conf : Alias_conf.t) =
-         ~targets:Alias
-         ~targets_dir:dir
-         ~scope)
+  let add_alias () =
```
I think this code would be slight clearer if written this way:

```ocaml
let enabled =
  match alias_conf.enabled_if with
  | None -> true
  | Some blang -> ...
in
if enabled then
  add_alias sctx ...
```

diml commented on src/stdune/sexp.ml:
```diff
@@ -81,6 +81,11 @@ module Of_sexp = struct
     Printf.ksprintf (fun msg ->
       of_sexp_error loc ?hint ("No variables allowed " ^ msg)) fmt
 
+  let _fix f =
```
Why not modify the type of `fix` instead of adding this function?

rgrinberg replied:
This is just an artefact from the rebase. Removed.

diml commented on src/jbuild.ml:
```diff
@@ -378,6 +378,42 @@ module Preprocess = struct
+           Compare (op, x, y))))
+    in
+    let t =
+    fix begin fun (t : String_with_vars.t Blang.t Sexp.Of_sexp.t) ->
```
Indentation looks wrong

diml commented on src/jbuild.ml:
```diff
@@ -378,6 +378,42 @@ module Preprocess = struct
+      [ "or", repeat t >>| (fun x -> Or x)
+      ; "and", repeat t >>| (fun x -> And x)
+      ] @ ops
+      |> sum ~rest:(String_with_vars.t >>| fun v -> Expr v)
```
I'm dubious about this `rest` argument. I feel like it's going to make error messages worse. Could you use `if_list` instead?

rgrinberg commented:
@diml I've addressed all your comments. Actually, `~rest` was only there because it's older than `if_list`. I've simply didn't reconsider the implementation using the new API after rebasing.

diml commented on doc/dune-files.rst:
```diff
@@ -824,6 +828,36 @@ doesn't start by `-`, you can simply quote it: ``("x" y z)``.
+
+   expr := (and <expr>+)
+         | (or <expr>+)
+         | (compare <op> <template> <template>)
```
should be: `(<op> <template> <template>)`

diml commented on src/value.ml:
```diff
@@ -1,18 +1,46 @@
+let compare x y =
+  match x, y with
+  | String x, String y -> String.compare x y
+  | Path x, Path y -> Path.compare x y
```
The `Dir x, Dir y` case is missing

rgrinberg replied:
Actually, we don't need this function anyway. I got rid of it.

rgrinberg commented:
@diml oops, the subsequent commit removed it during the rebase. The doc is now back.

rgrinberg commented:
Squashed & Rebased. Also added a change log entry.
